proofs/eo/cpc/Cpc.eo

```diff
@@ -367,6 +367,57 @@
   :conclusion (= a b)
 )
 
+
+; program: $get_annihilate
+; args:
+; - f (-> T U U): The type of function we are searching over, expected to be right-assoc-nil.
+; - t U: The term to compute whether it annilates.
+; - zero U: The zero element of f.
+; return: >
+;   zero, if it is found as a subterm of t beneath f-applications.
+(program $get_annihilate_rec ((T Type) (U Type) (f (-> T U U)) (a U) (b U) (zero U))
+  ((-> T U U) U U) U
+  (
+    (($get_annihilate_rec f zero zero)    zero)
+    (($get_annihilate_rec f (f a b) zero) (eo::ite (eo::is_eq ($get_annihilate_rec f a zero) zero)
+                                            zero
+                                            ($get_annihilate_rec f b zero)))
+  )
+)
+
+; program: $get_annihilate
+; args:
+; - t S: The term to compute whether it annilates.
+; return: >
+;   The zero element for t, if it is found.
+(define $get_annihilate ((S Type :implicit) (t S))
+  (eo::match ((T Type) (U Type) (V Type) (x V) (x1 T) (x2 U :list) (m Int)
+              (xb1 (BitVec m)) (xb2 (BitVec m) :list))
+    t
+    (
+    ((or x1 x2)        ($get_annihilate_rec or t true))
+    ((and x1 x2)       ($get_annihilate_rec and t false))
+    ((re.union x1 x2)  ($get_annihilate_rec re.union t re.all))
+    ((re.inter x1 x2)  ($get_annihilate_rec re.inter t re.none))
+    ((bvor xb1 xb2)    ($get_annihilate_rec bvor t ($bv_ones ($bv_bitwidth (eo::typeof xb1)))))
+    ((bvand xb1 xb2)   ($get_annihilate_rec bvand t (eo::to_bin ($bv_bitwidth (eo::typeof xb1)) 0)))
+    )
+  )
+)
+
+; rule: annihilate
+; implements: ProofRule::ANNIHILATE
+; args:
+; - eq Bool: The equality to be proven by this rule.
+; requires: >
+;   That t annihilates to a zero element, which is given by the right hand side of eq.
+; conclusion: The given equality.
+(declare-rule annihilate ((U Type) (t U) (zero U))
```

Comment:
The checking logic looks alright to me, but I would prefer it if we can simplify it to something that mimics what is done in cvc5:
 `:requires ((eo::and (eo::is_eq ($get_zero_for t) zero) ($contains t zero) true))`

Reply:
I refactored the signature now.